flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/gateway/context/SessionContextTest.java

```diff
@@ -126,61 +140,87 @@ public void testSetAndResetKeyInConfigOptions() throws Exception {
+        assertTrue(getConfiguration().getBoolean(OBJECT_REUSE));
     }
 
     @Test
-    public void testSetWithConfigOptionAndResetWithYamlKey() throws Exception {
-        SessionContext sessionContext = createSessionContext();
+    public void testSetWithConfigOptionAndResetWithYamlKey() {
         // runtime config option and has deprecated key
         sessionContext.set(TABLE_PLANNER.key(), "blink");
-        assertEquals(
-                "blink", getConfiguration(sessionContext).get(TABLE_PLANNER).name().toLowerCase());
+        assertEquals("blink", getConfiguration().get(TABLE_PLANNER).name().toLowerCase());
 
         sessionContext.reset(TABLE_PLANNER.key());
-        assertEquals(
-                "old", getConfiguration(sessionContext).get(TABLE_PLANNER).name().toLowerCase());
-        assertEquals(
-                "old", getConfigurationMap(sessionContext).get("execution.planner").toLowerCase());
+        assertEquals("old", getConfiguration().get(TABLE_PLANNER).name().toLowerCase());
+        assertEquals("old", getConfigurationMap().get("execution.planner").toLowerCase());
     }
 
     @Test
-    public void testSetAndResetKeyNotInYaml() throws Exception {
-        SessionContext sessionContext = createSessionContext();
+    public void testSetAndResetKeyNotInYaml() {
         // other property not in yaml and flink-conf
         sessionContext.set("aa", "11");
         sessionContext.set("bb", "22");
 
-        assertEquals("11", getConfigurationMap(sessionContext).get("aa"));
-        assertEquals("22", getConfigurationMap(sessionContext).get("bb"));
+        assertEquals("11", getConfigurationMap().get("aa"));
+        assertEquals("22", getConfigurationMap().get("bb"));
 
         sessionContext.reset("aa");
-        assertNull(getConfigurationMap(sessionContext).get("aa"));
-        assertEquals("22", getConfigurationMap(sessionContext).get("bb"));
+        assertNull(getConfigurationMap().get("aa"));
+        assertEquals("22", getConfigurationMap().get("bb"));
 
         sessionContext.reset("bb");
-        assertNull(getConfigurationMap(sessionContext).get("bb"));
+        assertNull(getConfigurationMap().get("bb"));
+    }
+
+    @Test
+    public void testAddJar() throws IOException {
+        sessionContext.addJar(udfJar.getPath());
+        assertEquals(
+                Collections.singletonList(udfJar.toURI().toURL().toString()),
+                getConfiguration().get(JARS));
+
+        // reset to the default classloader
+        sessionContext.reset();
+        assertEquals(
+                Collections.singletonList(udfJar.toURI().toURL().toString()),
+                getConfiguration().get(JARS));
+    }
+
+    @Test
+    public void testAddIllegalJar() {
+        validateAddJarException("/path/to/illegal.jar", "JAR file does not exist");
+    }
+
+    @Test
+    public void testRemoteJar() {
+        validateAddJarException("hdfs://remote:10080/remote.jar", "JAR file does not exist");
```

Comment:
I think this exception message is misleading, the jar actually exists there. We just doesn't support remote jar. 

We should check the protocal/schema of the url, we only support `file` now. You can use `URI.create(jarPath).getSchema()` to get the schema and throw a meaningful message if not `file`.